base_layer/core/src/validation/helpers.rs

```diff
@@ -21,43 +21,75 @@
+use tari_crypto::tari_utilities::{epoch_time::EpochTime, hash::Hashable, hex::Hex};
+
 pub const LOG_TARGET: &str = "c::val::helpers";
-use crate::proof_of_work::get_median_timestamp;
-use tari_crypto::tari_utilities::hex::Hex;
 
-/// This function tests that the block timestamp is greater than the median timestamp at the specified height.
-pub fn check_median_timestamp<B: BlockchainBackend>(
-    db: &B,
+/// This function tests that the block timestamp is less than the FTL
+pub fn check_timestamp_ftl(
     block_header: &BlockHeader,
-    height: u64,
-    rules: ConsensusManager,
+    consensus_manager: &ConsensusManager,
 ) -> Result<(), ValidationError>
 {
-    if block_header.height == 0 || rules.get_genesis_block_hash() == block_header.hash() {
-        return Ok(()); // Its the genesis block, so we dont have to check median
+    if block_header.timestamp > consensus_manager.consensus_constants(block_header.height).ftl() {
+        warn!(
+            target: LOG_TARGET,
+            "Invalid Future Time Limit on block:{}",
+            block_header.hash().to_hex()
+        );
+        return Err(ValidationError::BlockHeaderError(
+            BlockHeaderValidationError::InvalidTimestampFutureTimeLimit,
+        ));
     }
-    let min_height = height.saturating_sub(
-        rules
-            .consensus_constants(block_header.height)
-            .get_median_timestamp_count() as u64,
+    Ok(())
+}
+
+/// Returns the median timestamp for the provided timestamps.
+pub fn calc_median_timestamp(timestamps: &[EpochTime]) -> EpochTime {
+    assert_eq!(
+        timestamps.is_empty(),
+        false,
+        "calc_median_timestamp: timestamps cannot be empty"
     );
-    let timestamps = fetch_headers(db, min_height, height)?
-        .iter()
-        .map(|h| h.timestamp)
-        .collect::<Vec<_>>();
-    let median_timestamp = get_median_timestamp(timestamps).ok_or_else(|| {
-        error!(target: LOG_TARGET, "Validation could not get median timestamp");
-        ValidationError::BlockHeaderError(BlockHeaderValidationError::InvalidTimestamp)
-    })?;
+    trace!(
+        target: LOG_TARGET,
+        "Calculate the median timestamp from {} timestamps",
+        timestamps.len()
+    );
+
+    let mid_index = timestamps.len() / 2;
+    let median_timestamp = if timestamps.len() % 2 == 0 {
+        (timestamps[mid_index - 1] + timestamps[mid_index]) / 2
+    } else {
+        timestamps[mid_index]
+    };
+    trace!(target: LOG_TARGET, "Median timestamp:{}", median_timestamp);
+    median_timestamp
+}
 
+pub fn check_header_timestamp_greater_than_median(
+    block_header: &BlockHeader,
+    timestamps: &[EpochTime],
+) -> Result<(), ValidationError>
+{
+    if timestamps.is_empty() {
+        return Err(ValidationError::BlockHeaderError(
+            BlockHeaderValidationError::InvalidTimestamp,
+        ));
+    }
+
+    let median_timestamp = calc_median_timestamp(timestamps);
     if block_header.timestamp < median_timestamp {
```

Comment:
I believe this should be <=, not just <. 
If the block timestamp is allowed to be equal to the median. Then the block chain will never move forward in terms of timestamps.

Reply:
It sounds like that is possible after a few blocks with manipulated timestamps given enough mining control. Changing this shouldn't break consensus unless a node actually sets the same epoch time - I'll have to test if this is ok.

Reply:
I've made this change after testing it by synching a node from scratch - all blocks passed this validation